### ISSUE TITLE
Metrics by ensemble ID

### ISSUE THREAD

**Issue by mpinkerton-oasis** [OasisLMF/OasisPlatform#397]
Add the ability for custom models to output an ensemble file that will be used to create metrics (AAL and EP) by ensemble. It is assumed for now that each ensemble is equiprobable and has the same number of samples, though this will be be generally true. An structure of the new “ensemble_mapping.csv” file is:

sidx	ensemble_id	sample_id
1	1	1
2	2	1
3	3	1
4	4	1
5	5	1
6	6	1
7	7	1
8	8	1
9	9	1

- [x] Update aalcalc to calculate by ensemble if file is present.

- [x] Update leccalc to calculate by ensemble if file is present.

- [ ] Create note on how to configure ensemble output.

**Comment by hchagani-oasislmf** [comment]
In `ktools` branch `feature/metrics-by-ensemble-id`, `aalcalc` calculates by ensemble if file is present. Would be good for somebody from JBA to assess whether behaviour witnessed is desirable.

In order to facilitate this, the example ensemble file used `ensemble.csv`:

```
sidx,ensemble_id,summary_id
1,1,1
2,2,1
3,3,1
4,4,1
5,5,1
6,6,1
7,7,1
8,8,1
9,9,1
10,10,1
```
must be converted to binary format using the new executable `ensembletobin` (inverse operation can be performed by `ensembletocsv`):

```
# Creates ensemble.bin file
$ ensembletobin < ensemble.csv > ensemble.bin

# Creates ensemble.csv file
$ ensembletocsv < ensemble.bin > ensemble.csv
```

**Question for JBA:** is `summary_id` column required? Currently not using it in the code as the first two columns give all the information required. Perhaps there is something I am overlooking?

The results can be illustrated with `PiWind`. If `ensemble.bin` file is not in `static/` directory, the usual output is seen:
```
$ aalcalc -Kgul_S1_summaryaalcalc
summary_id,type,mean,standard_deviation,exposure_value
1,1,256798.940000,743558.938231,6800000.000000
1,2,257437.147558,764282.085048,6800000.000000
```
However, if `static/ensemble.bin` exists, the `ensemble_id` column is added:
```
$ aalcalc -Kgul_S1_summaryaalcalc
summary_id,type,mean,standard_deviation,exposure_value,ensemble_id
1,1,256798.940000,743558.938231,6800000.000000,0
1,2,257437.147558,764282.085048,6800000.000000,0
1,2,256945.304978,758580.863572,6800000.000000,1
1,2,261861.847367,769542.432956,6800000.000000,2
1,2,265180.586612,790800.958472,6800000.000000,3
1,2,250495.037375,761087.342188,6800000.000000,4
1,2,267851.532337,775116.764504,6800000.000000,5
1,2,253030.920375,746052.010748,6800000.000000,6
1,2,262449.068830,764902.442369,6800000.000000,7
1,2,256188.320086,775840.555262,6800000.000000,8
1,2,259650.463721,763736.892983,6800000.000000,9
1,2,240718.393896,738923.144969,6800000.000000,10
```
When `ensemble_id = 0`, the calculation has been performed across all ensembles.

**Comment by carlfischerjba** [comment]
Thanks for this @hchagani-oasislmf . Which files or streams would need to contain an `ensemble_id` field (apart from `ensemble.bin`)?

My colleagues may have more questions.

**Comment by carlfischerjba** [comment]
Or maybe I've misunderstood. What would the input to `aalcalc` look like in your example?

**Comment by hchagani-oasislmf** [comment]
No problem @carlfischerjba . Only `ensemble.bin` needs to contain `ensemble_id`. When executed, `aalcalc` looks for the `ensemble.bin` file in the `static` directory (which is where other model files such as `footprint.bin` and `vulnerability.bin` are located). If `static/ensemble.bin` is found, `aalcalc` outputs results across all ensembles (which it does already) and by ensemble.

So, the only input required is that `ensemble.bin` file. It should be in the following format:

```
sidx,ensemble_id,sample_id
1,1,1
2,2,1
3,3,1
4,4,1
5,5,1
6,6,1
7,7,1
8,8,1
9,9,1
10,10,1
```

The code doesn't use `sample_id` at the moment. It doesn't need to as it determines whether an `ensemble_id` is shared across other `sidx` as the file is read in. I'm not sure whether you need this column for anything, but I would suggest dropping it if you don't.

**Comment by DanielFEvans** [comment]
The sample_id is intended to be used in future. The idea is to split out variations in the model parameters (the ensemble IDs) and stochastic variations introduced to the model (the sample IDs). The former might be that we modify the parameters our flood modelling to predict greater or lesser flood severity on average, while the stochastic uncertainty would be random differences in exposure location, actual damage, and so on (which closely represents how most models use the sidx currently - random samples from fixed input distributions of vulnerability/hazard).

**Comment by DanielFEvans** [comment]
...which means that we envision one ensemble ID (set of input parameters) producing many sample IDs (the stochastic sampling). We might do nine ensemble samples, each with 100 stochastic samples.

**Comment by hchagani-oasislmf** [comment]
I see. So, if I've understood you correctly, rather than calculating by `ensemble_id`, you'd want to calculate by `ensemble_id`-`sample_id` combination?

**Comment by DanielFEvans** [comment]
For the outputs, we do want the results by `ensemble_id` as you showed, actually - but with the mean and standard deviation calculated across all `sample_id`s. You could think of the previous situation as being `sidx = sample_id`, and the unstated `ensemble_id` was always `1` - we now want an output for each `ensemble_id`.

The idea is that `ensemble_id` will represent something meaningful that the user is interested in, but `sample_id` is just a random sample that, on its own, isn't interesting.

**Comment by hchagani-oasislmf** [comment]
OK, I think I understand. Thanks. If correct, then the `ensemble.csv` file could look like this?:

```
sidx,ensemble_id,sample_id
1,1,1
2,1,2
3,1,3
4,2,1
5,2,3
6,2,2
7,3,3
8,3,2
9,3,1
10,4,1
```

In which case, it will calculate the mean and standard deviation for each `ensemble_id` across all `sample_id`s. If I use that as the `ensemble.bin` file, I get the following output:

```
$ aalcalc -Kgul_S1_summaryaalcalc
summary_id,type,mean,standard_deviation,exposure_value,ensemble_id
1,1,256798.940000,743558.938231,6800000.000000,0
1,2,257437.147558,764282.085048,6800000.000000,0
1,2,261329.246319,772840.054422,6800000.000000,1
1,2,257125.830029,760629.393707,6800000.000000,2
1,2,259429.284213,767927.392668,6800000.000000,3
1,2,240718.393896,738923.144969,6800000.000000,4
```

Is that format what you would expect?

**Comment by DanielFEvans** [comment]
Yep, that looks like what I'd expect.

**Comment by hchagani-oasislmf** [comment]
Changes merged into `develop` by PR https://github.com/OasisLMF/ktools/pull/140 and published in `ktools v3.3.4`. Outstanding issues:

1. Calculating losses by ensemble ID when period weights used (see new issue https://github.com/OasisLMF/ktools/issues/143).
2. Update documentation.
